python/lsst/daf/butler/core/dimensions/graph.py

```diff
@@ -0,0 +1,473 @@
+
+    def __repr__(self):
+        return f"DimensionGraph({self._dimensions}, joins={self._joins})"
+
+    @property
+    def universe(self):
+        """The graph of all dimensions compatible with self (`DimensionGraph`).
+
+        The universe of a `DimensionGraph` that is itself a universe is
+        ``self``.
+        """
+        return self._universe
+
+    @property
+    def asSet(self):
+        """A set view of the `Dimensions` in ``self`` (`DimensionSet`).
+        """
+        return self._dimensions
+
+    @property
+    def optionals(self):
+        """A set containing all (recursive) optional dependencies of the
+        dimensions in this graph that are not included in the graph itself
+        (`DimensionSet`).
+        """
+        everything = self.asSet.expanded(optional=True)
+        return everything - self.asSet
+
+    @property
+    def elements(self):
+        """A set containing both `Dimension` and `DimensionJoin` objects
+        (`DimensionSet`).
+        """
+        return self._elements
+
+    def joins(self, summaries=True):
+        """Return the `DimensionJoin` objects held by this graph.
+
+        Parameters
+        ----------
+        summaries : `bool`
+            If `False`, filter out joins that summarize any other joins being
+            returned (in most cases only the most precise join between a group
+            of dimensions is needed).
+
+        Returns
+        -------
+        joins : `DimensionSet`
+            The joins held by this graph, possibly filtered.
+        """
+        if summaries:
+            return self._joins
+        return DimensionSet(
+            self.universe,
+            [join.name for join in self._joins if join.summarizes.isdisjoint(self._joins)]
+        )
+
+    def __len__(self):
+        return len(self._dimensions)
+
+    def __iter__(self):
+        return iter(self._dimensions)
+
+    def __contains__(self, key):
+        return key in self._dimensions
+
+    def __getitem__(self, key):
+        return self._dimensions[key]
+
+    def get(self, key, default=None):
+        """Return the element with the given name, or ``default`` if it
+        does not exist.
+
+        ``key`` may also be a `Dimension`, in which case an equivalent
+        object will be returned.
+        """
+        return self._dimensions.get(key, default)
+
+    @property
+    def names(self):
+        """The names of all elements (`set`-like, immutable).
+
+        The order of the names is consistent with the iteration order of the
+        `DimensionSet` itself.
+        """
+        return self._dimensions.names
+
+    @property
+    def links(self):
+        """The names of all fields that uniquely identify these dimensions in
+        a data ID dict (`frozenset` of `str`).
+        """
+        return self._dimensions.links
+
+    def __hash__(self):
+        return hash(self._dimensions)
+
+    def __eq__(self, other):
+        return self._dimensions == other
```

Comment:
Does `DimensionSet` know how to compare itself with `DimensionGraph`?

Reply:
Yes; it calls `toNameSet` on the other operand, and hence it's comparable with anything that accepts.